### ISSUE TITLE
Disable certain messages by default

### ISSUE THREAD

**Issue by AWhetter** [PyCQA/pylint#3512]
Following on from the discussion on #746, this is a draft proposal to start disabling certain messages by default. Furthermore, to make it easier to run pylint side by side with flake8, we want to document which messages are duplicated in flake8 and possible make them easier to disable as a group.

The following table contains every pylint message excluding those from plugins and py3k checks. For each message it is noted whether it is a style check or a functional check (by functional I mean it validates some sort of run time behaviour of the code being linted or checks for programmer mistakes) or other (when it's technically a style check but I don't think we would want to disable it when a user opts out of style checks, or if the message doesn't fit in the other categories), whether it duplicates a message in flake8, and **my opinion** on whether I think it should be disabled by default so that we can use it as a discussion point. I normally prefer the linter to be quite strict so it might be worth gathering user feedback on this. Another option would be to give messages levels of strictness and you pick a level to lint at.

Message Name | Message ID | Style or Functional or Other | flake8 ID | Disable by Default | Notes 
--- | --- | --- | --- | --- | --- 
blacklisted-name | C0102 | S |  | N | 
invalid-name | C0103 | S |  | N | 
empty-docstring | C0112 | S |  | Y | 
unneeded-not | C0113 | O | F713 | N | 
missing-module-docstring | C0114 | S | "D100,D104" | Y | 
missing-class-docstring | C0115 | S | "D101,D106" | Y | 
missing-function-docstring | C0116 | S | "D102,D103,D105" | Y | 
singleton-comparison | C0121 | S |  | N | 
misplaced-comparison-constant | C0122 | S |  | N | 
unidiomatic-typecheck | C0123 | F |  | N | 
non-ascii-name | C0144 | F |  | N | 
consider-using-enumerate | C0200 | O |  | N | 
consider-iterating-dictionary | C0201 | O |  | N | 
bad-classmethod-argument | C0202 | O |  | N | 
bad-mcs-method-argument | C0203 | O |  | N | 
bad-mcs-classmethod-argument | C0204 | O |  | N | 
single-string-used-for-slots | C0205 | F |  | N | 
line-too-long | C0301 | S | E501 | N | 
too-many-lines | C0302 | S |  | N | 
trailing-whitespace | C0303 | S | "W291,W293" | N | 
missing-final-newline | C0304 | S | W292 | N | 
trailing-newlines | C0305 | S |  | N | 
multiple-statements | C0321 | O | E701 | N | 
superfluous-parens | C0325 | S |  | N | 
bad-whitespace | C0326 | S | "E201,E202,E225" | Y | 
mixed-line-endings | C0327 | O |  | N | 
unexpected-line-ending-format | C0328 | O |  | N | 
bad-continuation | C0330 | S | E129 | Y | 
wrong-spelling-in-comment | C0401 | O |  | Y | Sort of disabled by default already because it requires enchant
wrong-spelling-in-docstring | C0402 | O |  | Y | Sort of disabled by default already because it requires enchant
invalid-characters-in-docstring | C0403 | O |  | Y | Sort of disabled by default already because it requires enchant
multiple-imports | C0410 | O |  |  | 
wrong-import-order | C0411 | S | I202 |  | 
ungrouped-imports | C0412 | S |  |  | 
wrong-import-position | C0413 | S |  |  | 
useless-import-alias | C0414 | O |  |  | 
import-outside-toplevel | C0415 | O |  |  | 
len-as-condition | C1801 | O |  | N | 
syntax-error | E0001 | F |  | N | 
unrecognized-inline-option | E0011 | F |  | N | 
bad-option-value | E0012 | F |  | N | 
init-is-generator | E0100 | F |  | N | 
return-in-init | E0101 | F |  | N | 
function-redefined | E0102 | F |  | N | 
not-in-loop | E0103 | F |  | N | 
return-outside-function | E0104 | F |  | N | 
yield-outside-function | E0105 | F |  | N | 
nonexistent-operator | E0107 | F |  | N | 
duplicate-argument-name | E0108 | F |  | N | 
abstract-class-instantiated | E0110 | F |  | N | 
bad-reversed-sequence | E0111 | F |  | N | Check false positive rate?
too-many-star-expressions | E0112 | F |  | N | 
invalid-star-assignment-target | E0113 | F |  | N | 
star-needs-assignment-target | E0114 | F |  | N | 
nonlocal-and-global | E0115 | F |  | N | 
continue-in-finally | E0116 | F |  | N | 
nonlocal-without-binding | E0117 | F |  | N | 
used-prior-global-declaration | E0118 | F |  | N | 
misplaced-format-function | E0119 | F |  | N | 
method-hidden | E0202 | F |  | N | 
access-member-before-definition | E0203 | F |  | N | 
no-method-argument | E0211 | F |  | N | 
no-self-argument | E0213 | O |  | N | 
invalid-slots-object | E0236 | F |  | N | 
assigning-non-slot | E0237 | F |  | N | 
invalid-slots | E0238 | F |  | N | 
inherit-non-class | E0239 | F |  | N | 
inconsistent-mro | E0240 | F |  | N | 
duplicate-bases | E0241 | F |  | N | 
class-variable-slots-conflict | E0242 | F |  | N | 
non-iterator-returned | E0301 | F |  | N | 
unexpected-special-method-signature | E0302 | F |  | N | 
invalid-length-returned | E0303 | F |  | N | 
invalid-bool-returned | E0304 | F |  | N | 
invalid-index-returned | E0305 | F |  | N | 
invalid-repr-returned | E0306 | F |  | N | 
invalid-str-returned | E0307 | F |  | N | 
invalid-bytes-returned | E0308 | F |  | N | 
invalid-hash-returned | E0309 | F |  | N | 
invalid-length-hint-returned | E0310 | F |  | N | 
invalid-format-returned | E0311 | F |  | N | 
invalid-getnewargs-returned | E0312 | F |  | N | 
invalid-getnewargs-ex-returned | E0313 | F |  | N | 
import-error | E0401 | F |  | N | 
relative-beyond-top-level | E0402 | F |  | N | 
used-before-assignment | E0601 | F |  | N | 
undefined-variable | E0602 | F | F821 | N | 
undefined-all-variable | E0603 | F |  | N | 
invalid-all-object | E0604 | F |  | N | 
no-name-in-module | E0611 | F |  | N | 
unpacking-non-sequence | E0633 |  |  |  | 
bad-except-order | E0701 | F |  | N | 
raising-bad-type | E0702 | F |  | N | 
bad-exception-context | E0703 | F |  | N | 
misplaced-bare-raise | E0704 | F |  | N | 
raising-non-exception | E0710 | F |  | N | 
notimplemented-raised | E0711 | F |  | N | 
catching-non-exception | E0712 | F |  | N | 
bad-super-call | E1003 | F |  | N | 
no-member | E1101 | F |  | Y | Raises lots of false positives 
not-callable | E1102 | F |  | N | 
assignment-from-no-return | E1111 | O |  | N | 
no-value-for-parameter | E1120 | F |  | N | 
too-many-function-args | E1121 | F |  | N | 
unexpected-keyword-arg | E1123 | F |  | N | 
redundant-keyword-arg | E1124 | F |  | N | 
missing-kwoa | E1125 | F |  | N | 
invalid-sequence-index | E1126 | F |  | N | 
invalid-slice-index | E1127 | F |  | N | 
assignment-from-none | E1128 | F |  | N | 
not-context-manager | E1129 | F |  | N | 
invalid-unary-operand-type | E1130 | F |  | N | 
unsupported-binary-operation | E1131 | F |  | N | 
repeated-keyword | E1132 | F |  | N | 
not-an-iterable | E1133 | F |  | N | 
not-a-mapping | E1134 | F |  | N | 
unsupported-membership-test | E1135 | F |  | N | 
unsubscriptable-object | E1136 | F |  | N | 
unsupported-assignment-operation | E1137 | F |  | N | 
unsupported-delete-operation | E1138 | F |  | N | 
invalid-metaclass | E1139 | F |  | N | 
unhashable-dict-key | E1140 | F |  | N | 
dict-iter-missing-items | E1141 | F |  | N | 
logging-unsupported-format | E1200 | F |  | N | 
logging-format-truncated | E1201 | F |  | N | 
logging-too-many-args | E1205 | F |  | N | 
logging-too-few-args | E1206 | F |  | N | 
bad-format-character | E1300 | F |  | N | 
truncated-format-string | E1301 | F |  | N | 
mixed-format-string | E1302 | F |  | N | 
format-needs-mapping | E1303 | F |  | N | 
missing-format-string-key | E1304 | F |  | N | 
too-many-format-args | E1305 | F |  | N | 
too-few-format-args | E1306 | F |  | N | 
bad-string-format-type | E1307 | F |  | N | 
bad-str-strip-call | E1310 | F |  | N | 
invalid-envvar-value | E1507 | F |  | N | 
yield-inside-async-function | E1700 | F |  | N | 
not-async-context-manager | E1701 | F |  | N | 
fatal | F0001 | F |  | N | 
astroid-error | F0002 | F |  | N | 
parse-error | F0010 | F |  | N | 
method-check-failed | F0202 | F |  | N | 
raw-checker-failed | I0001 | F |  | Y | Raises lots of false positives without configuration
bad-inline-option | I0010 | F |  | N | 
locally-disabled | I0011 | F |  | N | 
file-ignored | I0013 | F |  | N | 
suppressed-message | I0020 | F |  | N | 
useless-suppression | I0021 | F |  | N | 
deprecated-pragma | I0022 | F |  | N | 
use-symbolic-message-instead | I0023 | F |  | N | 
c-extension-no-member | I1101 | F |  | Y | Raises lots of false positives without configuration
literal-comparison | R0123 | F |  | N | 
comparison-with-itself | R0124 | F |  | N | 
no-self-use | R0201 | F |  | N | 
no-classmethod-decorator | R0202 | O |  | N | 
no-staticmethod-decorator | R0203 | O |  | N | 
useless-object-inheritance | R0205 | O |  | N | "Worth documenting disabling this in a ""supporting Python 2"" section?"
property-with-parameters | R0206 | O |  | N | 
cyclic-import | R0401 | F |  | N | 
duplicate-code | R0801 | O |  | Y | I find this creates false positives (for things like import groups) more than it helps find duplicate code.
too-many-ancestors | R0901 | O |  | Y | "It's often too hard to fix this without making significant design changes |  which often is not possible."
too-many-instance-attributes | R0902 | O |  | Y | "It's often too hard to fix this without making significant design changes |  which often is not possible."
too-few-public-methods | R0903 | O |  | Y | I agree with this one but some find it opinionated.
too-many-public-methods | R0904 | O |  | Y | "It's often too hard to fix this without making significant design changes |  which often is not possible."
too-many-return-statements | R0911 | O |  | Y | I find this is often fixed by arbitrararily splitting up the function in more functions that do not add value.
too-many-branches | R0912 | O | C901 | Y | I find this is often fixed by arbitrararily splitting up the function in more functions that do not add value.
too-many-arguments | R0913 | O |  | Y | "It's often too hard to fix this without making significant design changes |  which often is not possible."
too-many-locals | R0914 | O |  | Y | 
too-many-statements | R0915 | O |  | Y | 
too-many-boolean-expressions | R0916 | O |  | Y | 
consider-merging-isinstance | R1701 | O |  | N | 
too-many-nested-blocks | R1702 | O |  | Y | 
simplifiable-if-statement | R1703 | O |  | N | 
redefined-argument-from-local | R1704 | F |  | N | 
no-else-return | R1705 | O |  | N | 
consider-using-ternary | R1706 | O |  | N | 
trailing-comma-tuple | R1707 | F |  | N | 
stop-iteration-return | R1708 | F |  | N | 
simplify-boolean-expression | R1709 | O |  | N | 
inconsistent-return-statements | R1710 | S |  | Y | 
useless-return | R1711 | S |  | N | 
consider-swap-variables | R1712 | O |  | N | 
consider-using-join | R1713 | O |  | N | 
consider-using-in | R1714 | O |  | N | 
consider-using-get | R1715 | O |  | N | 
chained-comparison | R1716 | O |  | N | 
consider-using-dict-comprehension | R1717 | O |  | N | 
consider-using-set-comprehension | R1718 | O |  | N | 
simplifiable-if-expression | R1719 | O |  | N | 
no-else-raise | R1720 | O |  | N | 
unnecessary-comprehension | R1721 | O |  | N | 
consider-using-sys-exit | R1722 | O |  | N | 
no-else-break | R1723 | O |  | N | 
no-else-continue | R1724 | O |  | N | 
else-if-used | R5501 | O |  | N | 
unreachable | W0101 | F |  | N | 
dangerous-default-value | W0102 | F |  | N | 
pointless-statement | W0104 | F |  | N | 
pointless-string-statement | W0105 | F |  | N | 
expression-not-assigned | W0106 | F |  | N | 
unnecessary-pass | W0107 | O |  | N | 
unnecessary-lambda | W0108 | F |  |  | 
duplicate-key | W0109 | F |  |  | 
assign-to-new-keyword | W0111 | F |  | N | 
useless-else-on-loop | W0120 | F |  | N | 
exec-used | W0122 | F |  | N | Arguably opinionated?
eval-used | W0123 | F |  | N | Arguably opinionated?
confusing-with-statement | W0124 | F |  | N | 
using-constant-test | W0125 | F |  | N | 
missing-parentheses-for-call-in-test | W0126 | F |  | N | 
self-assigning-variable | W0127 | F |  | N | 
redeclared-assigned-name | W0128 | F |  | N | 
assert-on-string-literal | W0129 | F |  | N | 
comparison-with-callable | W0143 | F |  | N | Check false positive rate?
lost-exception | W0150 | F |  | N | 
assert-on-tuple | W0199 | F |  | N | 
attribute-defined-outside-init | W0201 | F |  | N | 
bad-staticmethod-argument | W0211 | O |  | N | 
protected-access | W0212 | F |  | N | 
arguments-differ | W0221 | F |  | N | 
signature-differs | W0222 | F |  | N | 
abstract-method | W0223 | F |  | N | 
super-init-not-called | W0231 | F |  | N | 
no-init | W0232 | F |  | N | 
non-parent-init-called | W0233 | F |  | N | 
useless-super-delegation | W0235 | O |  | N | 
invalid-overridden-method | W0236 | F |  | N | 
unnecessary-semicolon | W0301 | S | E703 | N | 
bad-indentation | W0311 | S | E111 | Y | 
mixed-indentation | W0312 | S | W191 | N | This is a syntax error in Python 3. Can this even be raised now?
wildcard-import | W0401 | F |  | N | 
deprecated-module | W0402 | F |  | N | 
reimported | W0404 | F |  | N | 
import-self | W0406 | F |  | N | 
preferred-module | W0407 | F |  | N | 
misplaced-future | W0410 | F |  | N | 
fixme | W0511 | O |  | Y | 
global-variable-undefined | W0601 | F |  | N | 
global-variable-not-assigned | W0602 | O |  | N | 
global-statement | W0603 | O |  | Y | 
global-at-module-level | W0604 | F |  | Y | 
unused-import | W0611 | F | F401 | N | 
unused-variable | W0612 | F | F841 | N | 
unused-argument | W0613 | F |  | N | 
unused-wildcard-import | W0614 | F |  | N | 
redefined-outer-name | W0621 | F |  | N | 
redefined-builtin | W0622 | F |  | N | 
redefine-in-handler | W0623 | F |  | N | 
undefined-loop-variable | W0631 | F |  | N | 
unbalanced-tuple-unpacking | W0632 | F |  | N | 
cell-var-from-loop | W0640 | F |  | N | 
possibly-unused-variable | W0641 | F |  | N | 
self-cls-assignment | W0642 | F |  | N | 
bare-except | W0702 | F | E722 | N | 
broad-except | W0703 | F |  | N | 
duplicate-except | W0705 | F |  | N | 
try-except-raise | W0706 | F |  | N | 
binary-op-exception | W0711 | F |  | N | 
raising-format-tuple | W0715 | F |  | N | 
wrong-exception-operation | W0716 | F |  | N | 
keyword-arg-before-vararg | W1113 | F |  | N | 
arguments-out-of-order | W1114 | F |  | N | 
non-str-assignment-to-dunder-name | W1115 | F |  | N | 
isinstance-second-argument-not-valid-type | W1116 | F |  | N | 
logging-not-lazy | W1201 | F |  | N | 
logging-format-interpolation | W1202 | F |  | N | 
logging-fstring-interpolation | W1203 | F |  | N | 
bad-format-string-key | W1300 | F |  | N | 
unused-format-string-key | W1301 | F |  | N | 
bad-format-string | W1302 | F |  | N | 
missing-format-argument-key | W1303 | F |  | N | 
unused-format-string-argument | W1304 | F |  | N | 
format-combined-specification | W1305 | O |  | N | 
missing-format-attribute | W1306 | F |  | N | 
invalid-format-index | W1307 | F |  | N | 
duplicate-string-formatting-argument | W1308 | O |  | N | 
f-string-without-interpolation | W1309 | O |  | N | 
anomalous-backslash-in-string | W1401 | O |  | N | 
anomalous-unicode-escape-in-string | W1402 | O |  | N | 
implicit-str-concat | W1404 | O |  | N | 
inconsistent-quotes | W1405 | S |  | Y | 
bad-open-mode | W1501 | F |  | N | 
boolean-datetime | W1502 | F |  | N | 
redundant-unittest-assert | W1503 | F |  | N | 
deprecated-method | W1505 | F |  | N | 
bad-thread-instantiation | W1506 | F |  | N | 
shallow-copy-environ | W1507 | F |  | N | 
invalid-envvar-default | W1508 | F |  | N | 
subprocess-popen-preexec-fn | W1509 | F |  | N | 
subprocess-run-check | W1510 | O |  | N |

**Comment by Pierre-Sassoulas** [comment]
Great issue, I think we could add a column for "remove entirely". For example, I think that everything regarding formatting that can be done by black could be safely removed from the codebase. Black fix it automatically, are we ever going to want to report errors that can be fixed automatically by a tool? Someone that does not know about black could take a painfully long time to apply pep8 manually. At the very least we should tell users to install black to do it (maybe in the score to not spam it in every message). 

Also, everything that has false-positive should really be disabled until it's fixed, because it makes adoption harder for most people and because Pylint is not supposed to be a linter that annoys you :) 

I very much agree with the comment on the ticket you linked that we should create multiple configuration profiles. Maybe even disable message when we detect the presence of installed package (isort, flake8 and black, maybe others ?)

**Comment by AWhetter** [comment]
The automatic disabling based on what's installed is a good idea. I like that. We should implement a way of disabling that behaviour but that should be easy enough to do.

**Comment by PCManticore** [comment]
Hey @AWhetter This is a great list, thank you so much for working on it!

Here are some of my notes regarding checks that could be disabled by default or removed entirely:

* `bad-classmethod-argument` and `bad-mcs-classmethod-argument` Disable by default.
* `bad-whitespace` and `bad-continuation`
  Remove entirely. We always play catch up with PEP 8 to support these in various code contexts.
* `wrong-import-order` and related import order checks
   Disable by default. Requires `isort` to be up-to-date regarding to what is a third party import or not.

* `import-outside-toplevel`
This is actually useful, why do you think it should be disabled by default? In the codebases that I've worked with, it was always better to move the imports at the top, especially since a lot of them were stdlib ones.

* `import-error` Disable by default. Requires a much better import engine to figure out when something actually doesn't exist. Maybe it can be lumped together with `no-member`.
* `no-member` Agree that we should disable it by default as it has perhaps the most false positives in our checks.
* `raw-checker-failed` You marked as disabled by default, but why? It indicates to the user that a module could not be checked since it's not pure Python. 
* `too-few-*` and `too-many-*` With some of them I agree that they should be disabled by default, but `too-few-public-methods` usually indicates that a class can just be a function.
* `fixme` Maybe we should remove this entirely.
* `global-at-module-level` and `global-used` What's wrong with these ones? 
* `mixed-indentation` Can be removed entirely at this point.


To answer @Pierre-Sassoulas I don't believe we should remove the formatting category wholesale regardless if the community aggregated around the use of `black`.
There are still folks out there that would want `black` to be more configurable, especially since it has some dubious choices in terms of what kind of refactoring it yields in certain contexts.
If I'd remove something from the formatting checks it would definitely be `bad-whitespace` and `bad-indentation`. I believe they don't add that much value and they always require some fiddling around to support whatever PEP 8 suggests. Tools like `pycodestyle` and `black` should deal with this mess.

In the long term, a stages solution similar to the one suggested in #746 is definitely my preferred solution, but for now disabling or removing some of these checks works as well.

**Comment by Pierre-Sassoulas** [comment]
> but too-few-public-methods usually indicates that a class can just be a function.

When the framework expects a class to modify a base behavior (like Django for model's Metaclass) you can't do that. Or see [this answer on Stack with 30+ upvote that tells you to disable `too-few-public-methods` without thinking every time you inherit from a class](https://stackoverflow.com/a/36093075/2519059). I think there is a lot of false positive for this one.

I might have got a little carried away with removing all the formatting messages. Probably not everyone uses black+flake8+isort :) 

Also, I agree that the stages gamify the lint even more and that's probably a good thing to be able to do. But I also like the standard style proposed [here](https://github.com/PyCQA/pylint/issues/746#issuecomment-434503103). Because in an IDE you won't be able to do 4 steps (Unless pylint switch stage alone ? It would be confusing to have errors appear. Just imagine having a junior create a "core" error in order to make the next stage errors disappear.).

That would probably mean pylint is a lot harder to integrate and understand. I read a blog article aimed toward beginner that said, and I paraphrase, "pylint is useless without configuration, so just use flake8". To address this I think pylint should be easier to use for first-time users or if you don't even know you're using pylint because it's integrated in your IDE. The very knowledgable users are already tailoring it to their need so we can just keep taking their `pylintrc` into account. 

So being able to be able to say that you want to follow the "well-known company" style and have pylint behave accordingly would be a huge simplification. I also think the stages are great when you launch pylint yourself, but will be very hard to integrate in an IDE.

**Comment by AWhetter** [comment]
@PCManticore Thanks for checking through the list. I've incorporated some of your changes:

* `bad-classmethod-argument` and `bad-mcs-classmethod-argument` are disabled by default.

* `bad-whitespace` and `bad-continuation` we're doing in #3571 

* `wrong-import-order` and friends are disabled by default.

* `import-outside-toplevel` is enabled by default. I don't know why I put is as disabled.

* `import-error` is disabled by default.

* `raw-checker-failed` is enabled. I misunderstood what this check was doing.

* `too-few-public-methods`: As @Pierre-Sassoulas  mentioned, I know that for a long time it would get raised on a class that inherited from something else. But this case is fine because usually you're trying to override some behaviour of the base class. I don't know if this is still the case? If not then I agree that we should keep this enabled.

* `fixme`: I think this has value for some people but it's hard to say without usage metrics. I don't use it.

* `global-at-module-level` and `global-used` were disabled because I think globals are used often enough that some people find this check is an opinionated one. I don't like globals. I think they cause problems and I would turn this check back on. But if enough people find it annoying then I think it makes sense to make it an opt-in check.

* `mixed-indentation` is marked as removable.

> To address this I think pylint should be easier to use for first-time users

I agree with this whole heartedly.

**Comment by danrneal** [comment]
This list includes errors from [flake8-docstrings](https://github.com/pycqa/flake8-docstrings) and [flake8-import-order](https://github.com/PyCQA/flake8-import-order) so it should probably denote those and also add error codes from [pep8-naming](https://github.com/PyCQA/pep8-naming)

**Comment by danrneal** [comment]
In addition having an automatic checker that turns off duplicate errors from other install checkers would be awesome but in the meantime, just a page in the docs that shows which to turn off with each checker would be supremely helpful.

**Comment by danrneal** [comment]
Just to add to the discussion as to a sane / useful default configuration.  My current config has led to very few false positives and make pylint a joy for me.

--disable=design, format, bare-except, unneeded-not, undefined-variable, unused-import, unused-variable, missing-module-docstring, missing-class-docstring, missing-function-docstring, no-self-argument, invalid-name, bad-classmethod-argument, bad-mcs-classmethod-argument

Design section is false positives galore for me.... so disabling that got rid of most annoyances of pylint
Format section has been ceded to black
All the rest are duplicate messages with flake8 (mccabe, pep8-naming, and flake8-docstring plugins are enabled)

Since using this config, pylint has close to 100% rate of helpful and useful messages.

**Comment by danrneal** [comment]
It looks like all of the design section in the above table by @AWhetter was marked as "should be disabled by default".  Would that be a welcome PR?  Looks like a good first issue for me in pylint.  Or does more discussion need to happen slash I'm jumping the gun?

**Comment by psass-antidot** [comment]
I think a PR would be welcome, there seems to be a lot of messages where everyone agrees that it should be disabled by default. Thank you for being willing to contribute to Pylint :)

**Comment by Pierre-Sassoulas** [comment]
I think a PR would be welcome, there seems to be a lot of messages where everyone agrees that it should be disabled by default. Thank you for being willing to contribute to Pylint :)

**Comment by PCManticore** [comment]
@Pierre-Sassoulas I don't think we reached a consensus on disabling the design checkers just yet. There are still valuable checks in that category that ought to be enabled by default, and the most contentious ones were `too-few-public-methods` and those related to class hierarchy.

**Comment by danrneal** [comment]
@PCManticore No worries, I made a PR for if and when a consensus is reached on disabling the design checker.

Which design messages do you think merit being enabled by default? From what I've seen most people who are using them change the default parameters for pretty much all of them.  If the messages are that individualistic, I think it makes an argument for them to be enabled on an individual basis.

**Comment by danrneal** [comment]
Following up on this issue.  I noticed that in the OP by @AWhetter all the design checker messages are recommend to be disabled by default.  I created a PR to disable the category by default #3650 since I find pylint to be infinitely more sane this way.  @Pierre-Sassoulas approved my PR but wanted another reviewer before merging.  @PCManticore chimed in and didn't think we had a consensus to disable design by default yet.  So to spur some more conversation along these lines which of the design checks does anyone find should remain enabled? (I can look to modifying my PR to include those).

The messages in questions are: too-few-public-methods, too-many-public-methods, too-many-return-statements, too-many-branches, too-many-arguments, too-many-locals, too-many-statements

**Comment by Pierre-Sassoulas** [comment]
I might have been unclear, I think too-many-return-statements, too-many-branches, too-many-arguments, too-many-locals, too-many-statements are all useful indication of high level design flaw (maybe encompassed by the Mc Cabe complexity metric though),  too-many-public-methods as well. Only, too-few-public-methods is unavoidable if you use inheritance a lot, or if you use django (Meta in Model or Form) and I think this one should be disabled.

**Comment by danrneal** [comment]
Would you say that these checks are highly individualistic depending on the project?  It seems that it isn't possible to have good defaults since how many is too-many or too-few varies so much project to project.  If a user is going to configure these defaults, they would also be able to enable them.  However, for someone who is looking for a pip-and-play option, these messages can be a nuisance.  My view is without proper individual configuration these checks ought to be disabled.

**Comment by Pierre-Sassoulas** [comment]
I think in W1203 the f-string option should become the default. If you have so much logging that lazy formatting will make a difference, you can change the default ? See [this comment](https://github.com/PyCQA/pylint/issues/2354#issuecomment-414526879)

**Comment by thejcannon** [comment]
This issue has some age, but recommend adding a column for overlap with `mypy` as well!

**Comment by DanielNoord** [comment]
Let's make this actionable:
- [ ] Fix the default value of `disable` and `enable`
- [ ] Look into the list of to-be-removed messages from this post
- [ ] Create new issue to gather additional feedback on sane defaults. This allows closing this issue which has become daunting for anybody wanting to provide actual feedback.

Based on the above list and discussion the following messages would be disabled by default:
```python
missing-module-docstring,
missing-class-docstring,
missing-function-docstring,

too-many-ancestors,
too-many-instance-attributes,
too-few-public-methods,
too-many-public-methods,
too-many-return-statements,
too-many-branches,
too-many-arguments,
too-many-locals,
too-many-statements,
too-many-boolean-expressions,
too-many-nested-blocks,

bad-classmethod-argument,
bad-mcs-classmethod-argument,

duplicate-code,

fixme,

empty-docstring,
invalid-characters-in-docstring,

inconsistent-quotes,

bad-indentation,

global-statement,
global-at-module-level,

inconsistent-return-statements,

import-error,
no-member,
c-extension-no-member,

wrong-import-order,
ungrouped-imports,
wrong-import-position,
```

I have some comments:
`empty-docstring` should be enabled. I don't see why you would want empty docstrings, but perhaps I am missing something? Same goes for `invalid-characters-in-docstring`

`inconsistent-quotes` is personal preference but easily disabled and not really at risk of false-positives. I think we should keep it enabled by default.

`bad-indentation` is a valuable checker imo and should be enabled.

The messages about `globals` are also valid and not at risk of false-positives. 

I like `inconsistent-return-statements` so I would keep it in.

The `no-member` and `import-error` messages are indeed at risk of creating false positives. It would make sense to turn them of by default although on the other hand these are some of the most valuable linting warnings. I would be okay with keeping them on.

The `import-order` messages are fine imo. If they are raised it is a good sign that your `isort` installation is out of date or broken.

**Comment by martimlobao** [comment]
> The `no-member` and `import-error` messages are indeed at risk of creating false positives. It would make sense to turn them of by default although on the other hand these are some of the most valuable linting warnings. I would be okay with keeping them on.

+1 on keeping these on by default. These errors are the main reason I use pylint over flake8.

**Comment by DanielNoord** [comment]
I'd like to try and get this merged before `2.15`. Please see my previous comment, but if nobody else wants to exclude or include any messages in the default disables the list I propose is:

```
missing-module-docstring,
missing-class-docstring,
missing-function-docstring,

too-many-ancestors,
too-many-instance-attributes,
too-few-public-methods,
too-many-public-methods,
too-many-return-statements,
too-many-branches,
too-many-arguments,
too-many-locals,
too-many-statements,
too-many-boolean-expressions,
too-many-nested-blocks,

bad-classmethod-argument,
bad-mcs-classmethod-argument,

duplicate-code,

fixme
```

**Comment by Pierre-Sassoulas** [comment]
Regarding ``duplicate-code`` the original reasoning was:

> I find this creates false positives (for things like import groups) more than it helps find duplicate code.

Since then we added multiple option that alleviate this issue by permitting to ignore signatures, and imports. I think duplicate code can be a default check (granted the default value for those options is to ignore).

I would also keep ``too-many-statements``, if a file is 2000+ line long it deserve a refactor.

**Comment by Pierre-Sassoulas** [comment]
> I still think the checker is not robust enough. Remember how when we turned it on for pylint suddenly PRs failed for having duplicate-code even if they did not change anything in the files that were now flagged? For some we weren't even able to reproduce those warnings locally.

Yeah I can't argue with that, I just had an issue locally on the maintenance branch. 

> We have too-many-lines for this, right?

Sorry I've mistaken one for the other. Isn't 100+ lines for a function a little long and often easy to refactor ?

**Comment by Pierre-Sassoulas** [comment]
> The noice of these messages is probably also a reason not to adopt pylint as it requires initial tweaking instead of it running "nicely" out of the box.

I get your point, let's remove this too.

Although I want to point out that for legacy code it's expected that adopting pylint is not going to be a "pip and install" processes where you just have to follow the unicorns in a field of roses without thorns (yet).  In the documentation we suggest to use ``--errors-only`` then disable convention and refactor messages then disable only refactor message.  We don't have a lot of tool to guide user to the next step for code quality yet, so it could change later if we work on https://github.com/PyCQA/pylint/issues/5604.

So I think **right now** the default value should work for a new project that is just being created to make sure it does not get "out of control" (i.e. not with an old code base from an intern that is now a senior in mind).

**Comment by DanielNoord** [comment]
After discussion in #7111 it seems like just adding this list as a default value to `disable` won't work. Instead it is probably better to move the messages that we do not want to turn on by default to extensions.

**Comment by jacobtylerwalls** [comment]
If there are some messages that we don't want to demote as low as an extension (don't have any in mind, just saying...) then we could leave them enabled but then instead work on a "choose your config" tool as I think @Pierre-Sassoulas proposed elsewhere (e.g. defaults for ML/tensorflow, for Django/API dev etc.)

Just mentioning this to say it might end up we don't need to demote every message we identified on this thread to an extension. Each one should probably be its own PR.

**Comment by DanielNoord** [comment]
I'd say the `missing-docstring` messages could indeed be an extension. They seem like something you specifically want to turn on.

The `too-*` messages might actually be okay as is? I don't see that much benefit from moving them to extensions when users can also easily change the threshold values or disable single messages? 

The two `bad-*` messages can still become an extension.

`duplicate-code` should probably be an extension.

`fixme` should definitely be an extension. Or be removed. Who doesn't have todos?... But since it is already implemented, probably better to make it an extension.

Let me know what you think of those suggestions!

**Comment by jacobtylerwalls** [comment]
> The too-* messages might actually be okay as is? I don't see that much benefit from moving them to extensions when users can also easily change the threshold values or disable single messages?

Right, the biggest offender is `too-few-public-methods`, but we could just change the default min value from 2 to 0.

**Comment by Pierre-Sassoulas** [comment]
Moving checker to extension is not flexible at all : it creates a lot of work on our side and churn for users that have to learn about the fact that we moved something to extension (generally because they were disabling it already and we warn them that the disable is now useless which caused friction twice for a check they don't want). Or they need to to add the load-plugin if they wanted to have it. Just look at all the issue for ``no-self-use`` both on github and stackoverflow. This is going to happen again.

I think more generally we have a usability issue because currently users **needs to know about checkers which are internal to pylint**. Whereas what is really important for users imo is **messages being raised or not** on their code and the options to control them. Checkers are an abstraction that should not leak outside of pylint and we should not bother users with them except if their use of pylint is advanced enough that they are implementing their own checks. I think we did the right thing when we removed the section for each checker in the pylint configuration. I also think the right solution now is to have an internal default value for each messages that is modular from the way each message is implemented, and let the users disable or enable the message depending on what they think of pylint's default in their own configuration.

If we do that we can add a cool feature for external configuration templates to extends our configuration at the same time (like [eslint does it](https://github.com/dustinspecker/awesome-eslint#configs-by-well-known-companiesorganizations) with a final result as simple as: ``"extends": ["airbnb", "airbnb/hooks"]`` instead of copy pasting hundred of lines of configuration). 

Considering that, I think the best solution is to create option templates, and use enable/disable/load-plugin as option that extends the default values.

**Comment by DanielNoord** [comment]
> I also think the right solution now is to have an internal default value for each messages that is modular from the way each message is implemented, and let the users disable or enable the message depending on what they think of pylint's default in their own configuration.

The issue with that is that it would require even more changes from users. Anybody who would want any of the values we would start disabling by default would need to add them to `enable`.
_That's why the previous solution worked so well imo: it only requires changes for new users. Current users could update and keep using their current configuration without their CI failing._

> Considering that, I think the best solution is to create option templates, and use enable/disable/load-plugin as option that extends the default values.

Sure, but we're miles away from this. There are many things we need to consider before we can even start implementing this. I think we shouldn't link this issue to a project that nobody has shown interest in picking up in the near future.

If we agree on what I suggested above we only need 4 PRs:
`missing-*`, `bad-*`, `duplicate-code` and `fixme`.
We could even do an intermediate `2.15` release with only these changes so we have a very concise changelog which points to this important change.

**Comment by Pierre-Sassoulas** [comment]
> The issue with that is that it would require even more changes from users. Anybody who would want any of the values we would start disabling by default would need to add them to enable.

Sure, but:
- If we're making it optional it's probably not a really important message so it's more likely that it's a message users want to disable than to enable
- User had to upgrade ``load-plugin`` anyway ? They had to search for the name of the checker (``pylint.extensions.no_self_use``), it's not something you just know, unless you read the release notes or the [stackoverflow question](https://stackoverflow.com/a/72491225/2519059) that had 50 views per day for a month which imo is an indication of bad user experience.  Hell, I had to search myself right now. But if a user really like ``no-self-use`` they might now the name of the message because this is the concept they interact with when using pylint.
- it's better to be opt-in than opt-out, "having to enable explicitly if you really want to have it again or trusting the pylint teams to know what's important" is less annoying than "having to cleanup the configuration because of a message that you explicitly disabled already".

> That's why the previous solution worked so well imo: it only requires changes for new users. Current users could update and keep using their current configuration without their CI failing.

I don't understand this part, CI started to raise bad-option-value for old disable/enable for old users when it was moved to extension and we introduced ``unknown-option-value``. For new users there wasn't any problem as they just did not have no-self use ? Also if we disable some messages by default how are CI going to fail for old users ? There's just going to be less message raised ?

> Sure, but we're miles away from this. There are many things we need to consider before we can even start implementing this. I think we shouldn't link this issue to a project that nobody has shown interest in picking up in the near future.

I did not use eslint for a long time. It's just a design example based on something they did right and of what templates permits to do : A whole repository of configuration suited to everyone needs. Also let's imagine that we get an enormous backlash over disabling missing-docstring by default and want to revert the change. Do we want to add three values in a template's ``enable`` or do we want to move the checker from extension to standard checks again ?

**Comment by DanielNoord** [comment]
> * If we're making it optional it's probably not a really important message so it's more likely that it's a message users want to disable than to enable

I think `missing-function-docstring` is probably relied on a lot. I think it is an example of a message that should probably not be turned on by default, but for codebases that run `pylint` currently and expect it to be on it will be very annoying to see that is has now been disabled by default. Especially if they only notice a couple months after updating.

> * User had to upgrade `load-plugin` anyway ? [...]

Yeah, that's why I proposed to move a lower number of messages. By making a quick `2.15` release which is focused on this change we can also avoid some of the churn I think.

> * it's better to be opt-in than opt-out, "having to enable explicitly if you really want to have it again or trusting the pylint teams to know what's important" is less annoying than "having to cleanup the configuration because of a message that you explicitly disabled already".

I don't think that's true. Especially because you don't get a notification about the fact that something has now become opt-in. You could update to `2.15` and never notice that these messages have now been disabled and your CI would still pass.

> I don't understand this part, 

I meant that by adding these values to the default of `disable` any project which has its own `disable` config (probably almost all) won't notice any difference: they are already overwriting the default value. The only people that would "notice" the new default value of `disable` would be new users, that's why I liked that solution.

>  Do we want to add three values in a template's `enable` or do we want to move the checker from extension to standard checks again ?

Everybody agrees that having templates would be nice, but there is a reason why nobody has implemented them and other tools in the ecosystem also haven't. It's pretty difficult to allow both templates and normal configuration at the same time as you need to make very conscious decisions about what to parse in which order. With the `argparse` migration we tried to mirror the current situation and look how much we got wrong... Adding another type of configuration, parsing and decisions about what to parse when is difficult to get right.
I'm specifically thinking of https://github.com/PyCQA/pylint/issues/3696 and the added level of complexity templates would bring.

**Comment by Pierre-Sassoulas** [comment]
> >  I don't understand this part,

> I meant that by adding these values to the default of disable

Ha ok, I thought you were talking about what we did with ``no-self-use`` in 2.14. 

> The only people that would "notice" the new default value of disable would be new users, that's why I liked that solution.

Every users wanting to add "abstract-method" to the disable in CLI would notice that the default had 15 values because they would not be able to do ``--disable=abstract-method`` but would need to **type** ``--disable=abstract-method,missing-function-docstring... (10+ messages)...missing-module-docstring`` or have some messages suddenly appear. 

> By making a quick 2.15 release which is focused on this change we can also avoid some of the churn I think.

I have no issues about releasing 2.15 fast and releasing fast in general. 

My point was that we need to prevent users having to know the checker names (which they won't, and they can't as we'd be creating a new checker in extension) and be forced to open issues, read the doc or search on stackoverflow. I don't think a fast release would help with that. If we use the message symbol at least they have a chance to remember that the message exists because it has been for decades, literally.

>  You could update to 2.15 and never notice that these messages have now been disabled and your CI would still pass.

Yes that's the point, nothing to do when upgrading. It's true that it can be seen as silently "failing", but consider that we always favor false negatives vs false positives in pylint and also that those messages we're removing were not chosen at random. They are annoying and we collectively think that there is a good reason to not add them in pylint's default configuration. If the user did not explicitly enable them, then is is such a big deal if they are not enforced anymore ? It's not that different than having a new check appear when we add something to the default checker in a minor.

> I think missing-function-docstring is probably relied on a lot.

If ``missing-function-docstring`` is used a lot and I think it is, then we could keep it enabled by default. We should in fact. Being able to disable missing module/class docstring and not missing-function-docstring was the reason I started to contribute to pylint in the first place.

> you need to make very conscious decisions about what to parse in which order.
> I'm specifically thinking of https://github.com/PyCQA/pylint/issues/3696 and the added level of complexity templates would bring.

``enable``/``disable`` are a particular case because we need to upgrade the value a specific way not replace it (i.e. we extends the default and there is a special value "all" to handle).  but once we do it correctly for CLI, we can do it for template. Most other values are simply replaced. I'm thinking of "default pylint template", then "user's pylint configuration", then "Cli argument" for the order, do you think it make sense ? If there's a template in a configuration we load it first, if there is a pylint configuration linked in CLI with ``--rcfile`` we load it first,

The simple solution is to just hack it and extends the parsed CLI value with the messages we want to disable, there's no template to put into place, it's just the list of enable/disable where we add and remove values, but this is technical debt imo.

**Comment by DanielNoord** [comment]
> Every users wanting to add "abstract-method" to the disable in CLI would notice that the default had 15 values 

Yeah, but that is assuming that current users don't already have `disable=` in their configuration. I think it's very likely that current users already have those, so they wouldn't notice any difference if we provide a new default value.

But that point is moot, we have decided not to change the default already.

> I don't think a fast release would help with that.

I think we can avoid a lot of issues by making the changelog very clear and providing easy upgrade steps.

> They are annoying and we collectively think that there is a good reason to not add them in pylint's default configuration. If the user did not explicitly enable them, then is is such a big deal if they are not enforced anymore ? It's not that different than having a new check appear when we add something to the default checker in a minor.

Yes! This would be a very big deal for me. If you have become used to having `missing-function-docstring` on and no longer actively check PRs for docstrings since you now `pylint` does this it would be very annoying to notice 6 months down the road that this is no longer the case. If we decide to go that route I would consider that a breaking change that can't go in `2.x`.

> If `missing-function-docstring` is used a lot and I think it is, then we could keep it enabled by default. We should in fact. Being able to disable missing module/class docstring and not missing-function-docstring was the reason I started to contribute to pylint in the first place.

There is a difference between being used a lot and being a "sane default". I think it is a very nice check to have turned on, but should probably not be turned on by default as it obscures a lot of first-time runs.

> The simple solution is to just hack it and extends the parsed CLI value with the messages we want to disable, there's no template to put into place, it's just the list of enable/disable where we add and remove values, but this is technical debt imo.

But what if there is a template in the `cli` and there is a template in the `rcfile`? Which should go first? What if there are conflicting options in templates or configurations? These are all questions for the other issue but for me show that it is too big of an issue to let this issue depend on it.

**Comment by jacobtylerwalls** [comment]
The option I like the most (because I find it least surprising) is for the map of message states to have individual starting defaults for each message. `used-before-assignment` on by default. `fixme` off by default. Then no changes to the existing enable/disable system--the enable/disable system simply updates that message map after it is populated with its initial values. I want to bypass the discussion about enables/disables completely if we can!

@DanielNoord expresses concern about the backwards incompatibility of relying on a check like `missing-class-docstring` and having it silently go away. Would we really have a problem with doing this tomorrow with `fixme`? Some of the messages could be delayed to 3.0, but maybe we could just start with `fixme` and `duplicate-code`? I would want to add `useless-suppression` eventually.

We can
- make it REALLY obvious in the release notes
- log an informational message next to the score, and then remove it in 3.0. This doesn't help users of JSON reporters, but this gets most users:
```shell
------------------------------------------------------------------
Your code has been rated at 8.85/10 (previous run: 9.04/10, -0.19)

INFO: Some checks introduced before version 2.15 did not run and must be individually enabled if wanted
See https://github.com/...
```
***
What made `no-self-use` so painful was that we didn't think through the effect on `useless-suppression`. IIRC @cdce8p was surprised that it was on by default. We've already started weakening it at #5898. Maybe it should go off by default before anything else. After all, `# pylint: disable=fixme` is not really useless if the check would have run but for the default state in the message map which is out of the user's control. Useless suppression is supposed to help clean up your code base after pylint fixes false positives or you fix your code.

**Comment by DanielNoord** [comment]
> The option I like the most (because I find it least surprising) is for the map of message states to have individual starting defaults for each message. `used-before-assignment` on by default. `fixme` off by default. Then no changes to the existing enable/disable system--the enable/disable system simply updates that message map after it is populated with its initial values. I want to bypass the discussion about enables/disables completely if we can!

Agreed that this would be the most ideal solution.

> @DanielNoord expresses concern about the backwards incompatibility of relying on a check like `missing-class-docstring` and having it silently go away. Would we really have a problem with doing this tomorrow with `fixme`? Some of the messages could be delayed to 3.0, but maybe we could just start with `fixme` and `duplicate-code`? I would want to add `useless-suppression` eventually.

What about making this one of the changes that goes into `3.0` and then (sadly) blocking this issue until `3.0`? With that we can also remove the whole "official extensions" stuff. Every message that is now an extensions just gets moved into the "disabled by default" group and we no longer have to worry about moving stuff around.

**Comment by jacobtylerwalls** [comment]
> What about making this one of the changes that goes into 3.0 and then (sadly) blocking this issue until 3.0? With that we can also remove the whole "official extensions" stuff. Every message that is now an extensions just gets moved into the "disabled by default" group and we no longer have to worry about moving stuff around.

+1

Doesn't have to be sad, if we merge a great PR to the 3.0 branch it will increase the urgency of releasing the branch. :D

**Comment by DanielNoord** [comment]
By the way, if somebody is really interested in this I would totally support a ``--future-enable-disable`` flag that would allow testing this change in `2.x` 😄 

I would advise to create a new issue for this change though and keep this focused on what messages we want to disable and enable when a good mechanism for doing so is in place.

**Comment by Pierre-Sassoulas** [comment]
> The option I like the most (because I find it least surprising) is for the map of message states to have individual starting defaults for each message.

Sure ! The mapping can be a nice json or python dict:
```json
{
    "missing-function-docstring": true,
    "missing-module-docstring": false
}
```

But there's no cost to make this mapping look and work like a pylint configuration and load it before anything else (that's what I call a template). The following ini example contains the same information than the json/python dict:

```ini
enable=missing-function-docstring
disable=missing-module-docstring
```

Regarding @DanielNoord concern about message silently failing I just wanted to point out that we removed checks in minor previously, the big one being the py3k checkers (#4942) and ``bad-whitespace`` (#3577) but also lots of others that are listed in [``pylint.message._deleted_messages.py``](https://github.com/PyCQA/pylint/blob/01ea163c48d759f86d75dfb47e00796b7409abe1/pylint/message/_deleted_message_ids.py#L19).

I remember maybe seeing one or two issues opened asking where they went, and we actually REMOVED the message or the checker completely with no way to enable it again. I'm thinking that waiting for 3.0 is a really huge barrier to remove annoying message from pylint, especially when user can enable them again. It's also contrary to what we advertise in the [doc](https://pylint.pycqa.org/en/latest/user_guide/installation/upgrading_pylint.html#upgrading-pylint) 

> log an informational message next to the score, and then remove it in 3.0. This doesn't help users of JSON reporters, but this gets most users:

Would it work for you @DanielNoord ? We could disable the message if every message is disabled or enabled explicitely in the configuration.

**Comment by DanielNoord** [comment]
> I remember maybe seeing one or two issues opened asking where they went, and we actually REMOVED the message or the checker completely with no way to enable it again. I'm thinking that waiting for 3.0 is a really huge barrier to remove annoying message from pylint, especially when user can enable them again. It's also contrary to what we advertise in the [doc](https://pylint.pycqa.org/en/latest/user_guide/installation/upgrading_pylint.html#upgrading-pylint)

I know it's a bit arbitrary but for me there is a difference between removing a message (and probably documenting that in the changelog) than silently not running a message.

> Would it work for you @DanielNoord ? We could disable the message if every message is disabled or enabled explicitely in the configuration.

That just adds a lot of overhead as we would also need to handle cases where the messages were previously disabled anyway and when they are all enabled again.

Isn't working towards the ideal situation in `3.0` but already allowing that behind in flag in `2.x` and also get feedback on it better?

**Comment by jacobtylerwalls** [comment]
> The following ini example contains the same information than the json/python dict:

What I'm worried about is that if we call it "enable" and "disable" we start implicating questions of whether each layer overwrites or extends. Where as if we don't call it "enable" and "disable" we can't confuse it with the user-controlled enable/disables which only extend, not overwrite.

Enables/disables currently work by extending, i.e. `pylint mymodule --disable=fixme --disable=duplicate-code` disables both, and if that were to change, I would call it a regression. The reason I argued against the approach in #7111 was that it would introduce an "overwriting" paradigm into the existing "extension" paradigm, and I didn't want to change the status quo (or mix things up).

**Comment by Pierre-Sassoulas** [comment]
> What I'm worried about is that if we call it "enable" and "disable" we start implicating questions of whether each layer overwrites or extends

Right now if we have a checker in stdlib or an extension that is loaded and has say "example-message" in it, it's enabled. Then if we have a pylint configuration with ``disable=example-message`` we disable it, if the CLI has ``--enable=example-message`` we re-enable it (some trick with "all" and order of option might apply, but let's not worry too much about that*), if the file we're linting has ``pylint: disable=example-message`` we disable it for the file, a class inside the file has ``pylint: enable=example-message`` then we re-enable it for the class scope, and so on for function, and even for scope inside the function.

What I mean is that we already extends the value in the configuration using ini, in the CLI using an argparse option, in code using comment...  I don't think there's a question about how ``enable``/``disable`` works, it always extends the existing values, and the initial value is always everything enabled and nothing disabled (plugins can be loaded or not loaded, but if it's loaded  all messages are enabled).

So to implement this issue, we can change the fact that the original value is always True, or the fact that we always extends for enable/disable. Imo reusing the pylint configuration (well know format) and the extending mechanism (that is already everywhere) instead of adding yet another format and another mechanism is a good thing for long term maintenance. 

(*I'm pretty sure we do not use the same function for this everywhere, in particular I don't think we handle "all" comment in code disable, and the order of resolution is probably all other the place, but we could refactor this to have something generic and clear).

**Comment by jacobtylerwalls** [comment]
> What I mean is that we already extends the value in the configuration using ini, in the CLI using an argparse option, in code using comment... I don't think there's a question about how enable/disable works, it always extends the existing values, and the initial value is always everything enabled and nothing disabled (plugins can be loaded or not loaded, but if it's loaded all messages are enabled).

I think we're mostly on the same page, and I agree enable/disable is currently well understood. I'm just trying to express an agnosticism about whether the future "templating" issue has to be extends (i.e. has to be thought of enable/disable) or could be overwriting (and thus thought of as something else). I think it could go either way. However, if we make this equivalence in terminology right now:
```
this issue == templating == enable/disable
```
then that "locks us in" to extending rather than overwriting, because I completely agree that enable/disable is currently well understood and shouldn't change.

Forgive me if I'm repeating the obvious, I'm just trying to make sure I understand you. If you have The Washington Post's pylint template in your pylintrc but then use El País' pylint template at the command line, should that overwrite or extend? I'm not sure. Daniel doesn't sound sure either. I think if we call the templates "enable/disable templates" there's our answer--it has to extend. But I don't think it would violate "least surprise" if the El País template overwrote the Washington Post template. It's not a jinja2 template, it's more like "choose your starting point". (Hats off though to the person who decides to create nested pylint templates 😛 )

**Comment by Pierre-Sassoulas** [comment]
>  (Hats off though to the person who decides to create nested pylint templates 😛 )

Haha yes, also there's the per-directory configuration that is a whole project in itself and just another layer of extending/overwriting the conf where we'll have to take the same kind of decision (yet another step between CLI message control and per file message control).

I'm not suggesting to do the whole template feature for this, but loading something resembling a pylintrc and reusing disable/enable function (probably refactoring them for this) would prevent technical debt/complexity. I can do a MR in this sense if you want.

**Comment by DanielNoord** [comment]
> I'm not suggesting to do the whole template feature for this, but loading something resembling a pylintrc and reusing disable/enable function (probably refactoring them for this) would prevent technical debt/complexity. I can do a MR in this sense if you want.

I'd warn against any temporary functionality that will be superseded by full templating ability. That just adds more to deprecate at some point.

**Comment by Cheaterman** [comment]
Hi everyone and thanks for the hard work! Sorry for the noise, but what's the status on this? The thread is getting a bit hard to follow and I wonder if some parts of it are ready to be tested, or if users are currently still expected to configure the relaxed settings themselves? Thanks in advance!

**Comment by Pierre-Sassoulas** [comment]
We made no-self-use an extension in 2.14 (?), but otherwise this is still a work in progress. There's an actionable list here: https://github.com/pylint-dev/pylint/issues/3512#issuecomment-1118331249, but we discussed a lot after that, for duplicate code, what to move into extension or not... We probably need a maintainer to summarize what need to be done if we want to make this actionable again. There's also been evolution in what pylint can do (auto upgrade of config would be nice to have before #9088, #5462 )

**Comment by ShahidSandhu** [comment]
# pylint: skip-file
ignore=<default list>,<E501,W291,E999>

this code for ignoring error message of pylint is creating further error with E999 "./server/djangoapp/populate.py:2:9: E999 SyntaxError: invalid syntax". CAN ANY BODY SOLVE THIS.

**Comment by clauspruefer** [comment]
In contrast to the list `C0114`, `C0115` and `C0116` seem to be enabled  on GitHub actions (pylint version: `pylint-3.3.3-py3-none-any.whl` and `CPython (3.10.16)`.

Is this correct behaviour?

**Comment by jacobtylerwalls** [comment]
It's correct, because the list in this issue is still in a proposal stage. It's an accepted proposal, it just needs funding/volunteers.

**Comment by clauspruefer** [comment]
> It's correct, because the list in this issue is still in a proposal stage. It's an accepted proposal, it just needs funding/volunteers.

Thanks for the info. The C0114, C0115, C0116 like proposed should be implemented as fast as possible. The more default switches i have to disable in an GitHub automation process (which is currently the case), the more work i have to do.

And what is scientifically proven is that humans are lazy ...

**Comment by jacobtylerwalls** [comment]
Replying to https://github.com/pylint-dev/pylint/issues/3512#issuecomment-1172911486:

> Considering that, I think the best solution is to create option templates, and use enable/disable/load-plugin as option that extends the default values.

And https://github.com/pylint-dev/pylint/issues/3512#issuecomment-1172912436:

> Sure, but we're miles away from this. There are many things we need to consider before we can even start implementing this. I think we shouldn't link this issue to a project that nobody has shown interest in picking up in the near future.

I think this is the right design, and not such a hard lift. Here's a quick proof of concept: #10619

**Comment by DanielNoord** [comment]
> I think this is the right design, and not such a hard lift. Here's a quick proof of concept: [#10619](https://github.com/pylint-dev/pylint/pull/10619)

I still believe you and I both made valid arguments years ago that haven't been resolved yet.

Concretely I think the open questions are:
- How do we handle edge cases when specyfing multiple templates via different sources
- How do `enable`/`disable` work with respect to the templates, especially considering the order of arguments in both configuration files and the CLI matters for their interpretation

Somewhat related, but not critical to me as we never got to rounding off that problem/feature:
- How does templating work in a future where we have per-directory configs, which introduce another source of potentially conflicting configurations

Therefore, I still think the following solutions will work best in the long term if we can't solve the issues described above:
> The option I like the most (because I find it least surprising) is for the map of message states to have individual starting defaults for each message.

It introduces fewer footguns/edge cases and fits better with the needs I have seen. With the latter I'm referring to the fact that I have mostly seen people/contributors request to turn off certain messages by default, not to construct elaborate combinations of enable and disable patterns into templates.

**Comment by jacobtylerwalls** [comment]
> How do we handle edge cases when specyfing multiple templates via different sources
> How do enable/disable work with respect to the templates, especially considering the order of arguments in both configuration files and the CLI matters for their interpretation

It deserves a test, and I'm not making light of the testing effort that will be needed, but I think we just simply commit to a certain order. The linked sketch just processes `--message-sets` in the given order, before additional `--enable` and `--disable`, which is easy to reason about.

> With the latter I'm referring to the fact that I have mostly seen people/contributors request to turn off certain messages by default, not to construct elaborate combinations of enable and disable patterns into templates.

I think I agree with you: that's why pylint itself would "turn off certain messages by default" by inserting the list of messages extracted from whatever the consensus was (`fixme`, etc.) into [this list](https://github.com/pylint-dev/pylint/pull/10619/files#r2404677568). Users would not be constructing elaborate combinations, just plugin authors (potentially).

**Comment by DanielNoord** [comment]
> It deserves a test, and I'm not making light of the testing effort that will be needed, but I think we just simply commit to a certain order. The linked sketch just processes `--message-sets` in the given order, before additional `--enable` and `--disable`, which is easy to reason about.

Yeah it seems fair, but also something that can easily "go wrong" in configuration files. People tend to put options in alphabetical order (at least in my experience) and `enable` is likely to be parsed before any `disables` in `message-sets`.

We could also consider "storing" whether a message was intentionally flipped to off/on and warn the user if the `message-sets` is overwriting that? Or better yet, raise an error if `message-sets` is not the first option to be parsed?
That still doesn't solve the issue of overwrites from the CLI or per-directory configs but that is very finnicky anyway.
 
> > With the latter I'm referring to the fact that I have mostly seen people/contributors request to turn off certain messages by default, not to construct elaborate combinations of enable and disable patterns into templates.
> 
> I think I agree with you: that's why pylint itself would "turn off certain messages by default" by inserting the list of messages extracted from whatever the consensus was (`fixme`, etc.) into [this list](https://github.com/pylint-dev/pylint/pull/10619/files#r2404677568). Users would not be constructing elaborate combinations, just plugin authors (potentially).

As I see it, the only real benefit of the templating is when there are more than the 2 templates we will likely implement first (one being "everything is enabled" and the other "sane defaults") . Because if it is just 2 it would likely be simpler to just turn a couple of message off by default in the next major release.

I personally don't see us maintaining more than 2 templates in `pylint` itself, so it would come down to users and plugin authors. I also agree that users likely wouldn't be making these combinations so if I'm correct that we wouldn't be maintaining more than 2 templates the main "audience" for the API would be plugin authors.
For them the situation might be worse than it was before? Because now their messages would not be turned on by default, but every user will also need to explicitly define each message in the `enable` or add the revenat `message-set` as provided by the plugin. Something that for older, less maintained plugins might never become a thing?

----

The issue with plugins and their message sets also affects our extensions. Do we auto-load/enable all messages in a extension if it is enabled? Or should users enable both the extension and import the message set provided by the extensions? Or should the message set also include an attribute for extensions so that they can define which extensions they implicitly require?

**Comment by jacobtylerwalls** [comment]
> As I see it, the only real benefit of the templating is when there are more than the 2 templates we will likely implement first (one being "everything is enabled" and the other "sane defaults") . Because if it is just 2 it would likely be simpler to just turn a couple of message off by default in the next major release.

Another benefit of the templating is handing control to the users. If we flip the bit on the default message states for fixme and 10 others, but a user wants those eleven messages, how do they opt back in? With this, it's just `message-sets=pylint.core` or whatever we call it.

**Comment by jacobtylerwalls** [comment]
> The issue with plugins and their message sets also affects our extensions. Do we auto-load/enable all messages in a extension if it is enabled? Or should users enable both the extension and import the message set provided by the extensions? Or should the message set also include an attribute for extensions so that they can define which extensions they implicitly require?

I would start small, and leave whatever current behavior we have if you `--enable=plugin.foo` without loading the foo plugin.

**Comment by DanielNoord** [comment]
> Another benefit of the templating is handing control to the users. If we flip the bit on the default message states for fixme and 10 others, but a user wants those eleven messages, how do they opt back in? With this, it's just `message-sets=pylint.core` or whatever we call it.

But that's a one time migration thing, right? And if the "statistics" in this issue are correct, most people have these messages in their `disable` anyway 😄 But yeah, it is some migration effort for users, I'm aware of that.

> I would start small, and leave whatever current behavior we have if you `--enable=plugin.foo` without loading the foo plugin.

But that would mean we would handle messages in `plugins` differently than those in the base checkers, right? As those in base checkers are only enabeld if their are in `message-sets` or explicitly enabled, whereas messages in plugins are always enabled.

----

That has gotten me thinking: perhaps I misunderstood the proposal. I assumed that with `message-sets` no message would be enabled if it wasn't in a `message-set`, but that is of course not necessarily the case. And therefore plugins don't need to be special cases.
I guess `pylint.recommended` would have `all` in its `enable` and then `disable` the more opionated checks?
But then you get weird behaviour if the setting is `message-set=pylint_django.base,pylint.recommended` because if `all` is in `pylint.recommended` all `disables` in `pylint_django.base` might be undone. And if `all` isn't in `recommended` we would need to add all messages to all `messages-sets` and keep them in sync? And special case plugins.

-----

It is clearly too late for me to be thinking about this as I'm getting confused about the proposal. I see the benefit, but can't wrap my head around some of the edge cases. I hope we can iron those out, but from this little rambling of 40 minutes I'm wondering if we'd ever get to a system that is intuitive enough to all users/plugin authors and without too many footguns.

I'm also wondering if it is feasible to do the "turned off by default" for `4.0` and integrate that into templating for `5.0`? Like you said, in the end I think we all just want to make progress to "getting sane".

**Comment by jacobtylerwalls** [comment]
> I guess pylint.recommended would have all in its enable and then disable the more opionated checks?
But then you get weird behaviour if the setting is message-set=pylint_django.base,pylint.recommended because if all is in pylint.recommended all disables in pylint_django.base might be undone. And if all isn't in recommended we would need to add all messages to all messages-sets and keep them in sync? And special case plugins.

Close, but not quite what I had in mind. We would probably raise if you tried to provide an `"all"` anywhere in the message set.

> I'm also wondering if it is feasible to do the "turned off by default" for 4.0 and integrate that into templating for 5.0? Like you said, in the end I think we all just want to make progress to "getting sane".

We can explore this. We can also say this is experimental in 4.0 and say the only thing that's stable is if you add `--message-sets=pylint.recommended`, i.e. that exact value and nothing else, and then if we change the design we leave that as a shim for whatever is the "sane default" on the new api.

**Comment by Pierre-Sassoulas** [comment]
I reviewed the POC before reading the latest message here. Enabling or disabling extension or checker is going to be necessary to provide default/recommended value or we're going to either activate extension because their message are enabled or fail if the extension is not provided. Based on that, I think we're going to have (and it's not impossible to make) a 'style' object that is in fact a full configuration (minus rcfile and recursive option like 'all' in enable/disable that would send us all into madness) instead of a list of message so it's possible for exemple to define what are the regexes for invalid-name in one particular style.

```python
class Legacy:
    enable: list[Str] | none= ...
    disable: list[Str] | none = ...
    ...
    class_rgx: Str | none = ...
```

**Comment by DanielNoord** [comment]
> Close, but not quite what I had in mind. We would probably raise if you tried to provide an `"all"` anywhere in the message set.

So we would to add all messages to the `enable` explicitly?
 
> We can explore this. We can also say this is experimental in 4.0 and say the only thing that's stable is if you add `--message-sets=pylint.recommended`, i.e. that exact value and nothing else, and then if we change the design we leave that as a shim for whatever is the "sane default" on the new api.

Wouldn't it be simpler to do some disabling in `4.0` and explore messages sets for `5.0`?

> Based on that, I think we're going to have (and it's not impossible to make) a 'style' object that is in fact a full configuration (minus rcfile and recursive option like 'all' in enable/disable that would send us all into madness)

This is also what I'm a bit afraid of and which makes me even more scared to commit to templating. It seems a very big beast to tackle to just to disable a couple of opinionated messages by default, which I still believe is the main ask in this issue.

**Comment by jacobtylerwalls** [comment]
> So we would to add all messages to the enable explicitly?

It depends. We have to determine what the zero state is before config is applied: all on or all off. I was assuming to keep "all on" like today. Then there is less maintenance burden on the enables as checks are added.

**Comment by jacobtylerwalls** [comment]
> Wouldn't it be simpler to do some disabling in 4.0 and explore messages sets for 5.0?

Simpler for us maybe but not for upgrading users that get 12 messages turned off by default and want to know what to paste to get them back.

I'm fine waiting for 4.1 to do anything, just wanted to get a draft out.

**Comment by jacobtylerwalls** [comment]
Templating/message sets could start in 4.1 on opt-in basis, and then we delay cutting existing projects over to pylint.recommended until 5. (?)

**Comment by Pierre-Sassoulas** [comment]
I don't think we have to rush message sets or templating. If we want something simple to release 4.0 soon and focus on the task at hand in this issue we can add ``{"default_enabled": False,},`` in a bunch of messages we want to disable by default. (5mn task once we know the message we want to disable) And possibly add the breaking changes in the data structure to match that and be done with it (10mn task?). I'm implementing the upgrade configuration script right now to make this simple change be more user friendly but we could in theory do the same thing we did since 2.14 and just release new default without making a big deal about it.

Message sets are nice too (shorthand to keep the default from another pylint version will make upgrade easier), but I feel that if we add message sets it won't work well with templating later (the features are not perpendicular enough to both make sense at the same time). Imo templates is "the full feature" that we'll want in the long term. We have removed extensions or extensions moved to checker in the breaking changes, a simple message sets will not permit to take that into account easily as we also need the content of "load-plugins" defined.

Also if we choose 'all off' it's going to put a massive burden on users to make them "maintain their pylint configuration" and the old message sets are going to be deprecated all the time (i.e. 'core' can't exist because it's in fact 'core-pylint-4.0' ?). It's easy to upgrade pylint and see that there's new messages that are raised (the 'core' template has new messages). It's harder to check the documentation for all possible new messages and how to activate them (i.e. 'core' need to be upgraded to 'core-pylint-5.0' and you need to read the release notes to know that), which is kinda why we put opinionated things in extensions where it's hard to activate and we have ``--enable-all-extensions`` that is some kind of hidden template with ``load-plugins`` set to 'all on'.

**Comment by DanielNoord** [comment]
I agree with @Pierre-Sassoulas I think.

We can just provide the snippet to put in your `enable` for all the messages we disabled by default in the breaking changes setting of the changelog? That should cover most use cases?

**Comment by jacobtylerwalls** [comment]
Sounds good.
